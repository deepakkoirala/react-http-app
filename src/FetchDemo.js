import React from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';

class FetchDemo extends React.Component {
  static propTypes = {
    subreddit: PropTypes.string.isRequired
  }

  // Set up the default state with a property intitializer
  // instead of writing a whole constructor just for this
  // You can reference 'this.props' here if you need to.
  state = {
    posts: [],
    loading: true,
    error: null
  }

  componentDidMount() {
    // Remove the 'www.' to cause a CORS error (and see the error state)
    axios.get(`http://www.reddit.com/r/${this.props.subreddit}.json`)
      .then(res => {
        // Transform the raw data by extracting the nested posts
        const posts = res.data.data.children.map(obj => obj.data);

        // Update state to trigger a re-render.
        // Clear any errors, and turn off the loading indiciator.
        this.setState({
          posts,
          loading: false,
          error: null
        });
      })
      .catch(err => {
        // Something went wrong. Save the error in state and re-render.
        this.setState({
          loading: false,
          error: err
        });
      });
  }

  renderLoading() {
    return <div>Loading...</div>;
  }

  renderError() {
    return (
      <div>
        Something went wrong: {this.state.error.message}
      </div>
    );
  }

  renderPosts() {
    // Using destructuring to extract the 'error' and 'posts'
    // keys from state. This saves having to write "this.state.X" everwhere.
    const { error, posts } = this.state;

    if(error) {
      return this.renderError();
    }

    return (
      <ul>
        {posts.map(post =>
          <li key={post.id}>
            <span className="score">{post.score}</span>
            {post.title}
          </li>
        )}
      </ul>
    );
  }

  render() {
    const { subreddit } = this.props;
    const { loading } = this.state;

    return (
      <div>
        <h1>{`/r/${subreddit}`}</h1>
        {loading ? this.renderLoading() : this.renderPosts()}
      </div>
    );
  }
}


export default FetchDemo;
