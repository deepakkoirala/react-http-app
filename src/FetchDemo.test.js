import React from 'react';
import ReactDOM from 'react-dom';
import FetchDemo from './FetchDemo';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json'
import axios from 'axios';

it('renders the loading state', () => {
  const wrapper = shallow(
    <FetchDemo subreddit='reactjs'/>
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders the error state', () => {
  const wrapper = shallow(<FetchDemo subreddit='reactjs'/>);
  wrapper.setState({
    loading: false,
    error: {message: 'oops'}
  });
  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders the posts', () => {
  const wrapper = shallow(<FetchDemo subreddit='reactjs'/>);
  wrapper.setState({
    loading: false,
    posts: [
      {id: 1, score: 3, title: 'Best Post Ever'},
      {id: 2, score: 17, title: 'Second Best Post Ever'},
    ]
  });
  expect(toJson(wrapper)).toMatchSnapshot();
});

describe('fetching data', () => {
  const response = {
    data: { data: { children: [
      {data: {id: 1, title: "foo"}},
      {data: {id: 2, title: "bar"}}
    ]}}
  };

  it('fetches the data', () => {
    // Overwrite the axios.get function with a mock
    axios.get = jest.fn(() => Promise.resolve(response));

    // Must use "mount" instead of "shallow", otherwise componentDidMount()
    // won't be called.
    mount(<FetchDemo subreddit='reactjs'/>);

    // Check that axios.get was called correctly
    expect(axios.get).toBeCalledWith('http://www.reddit.com/r/reactjs.json');
  });

  // Pass the "done" param to this test because it
  // is testing async functionality
  it('re-renders with the data', (done) => {
    // Resolve the call with data
    axios.get = jest.fn(() => Promise.resolve(response));

    let wrapper = mount(<FetchDemo subreddit='reactjs'/>);

    // First time should say "Loading..."
    expect(toJson(wrapper)).toMatchSnapshot();

    try {
      setTimeout(() => {
        // After a timeout the promise will be resolved,
        // and the component's state will be updated
        expect(toJson(wrapper)).toMatchSnapshot();

        // Must call done() to stop the test
        done();
      }, 0);
    } catch(e) {
      done.fail(e);
    }
  });

  // Pass the "done" param to this test because it
  // needs to test async functionality
  it('re-renders with an error', (done) => {
    const error = {
      message: "End of the world"
    };

    // Reject the call with an error
    axios.get = jest.fn(() => Promise.reject(error));

    let wrapper = mount(<FetchDemo subreddit='reactjs'/>);

    // First time should say "Loading..."
    expect(toJson(wrapper)).toMatchSnapshot();

    try {
      setTimeout(() => {
        // After a timeout the promise will be resolved,
        // and the component's state will be updated
        expect(toJson(wrapper)).toMatchSnapshot();

        // Must call done()
        done();
      }, 0);
    } catch(e) {
      done.fail(e);
    }
  });
});

