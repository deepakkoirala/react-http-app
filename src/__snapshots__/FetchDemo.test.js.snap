exports[`fetching data re-renders with an error 1`] = `
<FetchDemo
  subreddit="reactjs">
  <div>
    <h1>
      /r/reactjs
    </h1>
    <div>
      Loading...
    </div>
  </div>
</FetchDemo>
`;

exports[`fetching data re-renders with an error 2`] = `
<FetchDemo
  subreddit="reactjs">
  <div>
    <h1>
      /r/reactjs
    </h1>
    <div>
      Something went wrong: 
      End of the world
    </div>
  </div>
</FetchDemo>
`;

exports[`fetching data re-renders with the data 1`] = `
<FetchDemo
  subreddit="reactjs">
  <div>
    <h1>
      /r/reactjs
    </h1>
    <div>
      Loading...
    </div>
  </div>
</FetchDemo>
`;

exports[`fetching data re-renders with the data 2`] = `
<FetchDemo
  subreddit="reactjs">
  <div>
    <h1>
      /r/reactjs
    </h1>
    <ul>
      <li>
        <span
          className="score" />
        foo
      </li>
      <li>
        <span
          className="score" />
        bar
      </li>
    </ul>
  </div>
</FetchDemo>
`;

exports[`test renders the error state 1`] = `
<div>
  <h1>
    /r/reactjs
  </h1>
  <div>
    Something went wrong: 
    oops
  </div>
</div>
`;

exports[`test renders the loading state 1`] = `
<div>
  <h1>
    /r/reactjs
  </h1>
  <div>
    Loading...
  </div>
</div>
`;

exports[`test renders the posts 1`] = `
<div>
  <h1>
    /r/reactjs
  </h1>
  <ul>
    <li>
      <span
        className="score">
        3
      </span>
      Best Post Ever
    </li>
    <li>
      <span
        className="score">
        17
      </span>
      Second Best Post Ever
    </li>
  </ul>
</div>
`;
